rfcs/45-stageless.md

```diff
@@ -0,0 +1,545 @@
+   if_needed_ordering_after: Vec<OrderingTarget>,
+   atomic_ordering_after: Vec<OrderingTarget>,
+   // The `TypeId` here is the type id of the flushing system
+   flushed_ordering_before: Vec<(TypeId, OrderingTarget)>,
+   flushed_ordering_after: Vec<(TypeId, OrderingTarget)>,
+   run_criteria: Vec<SystemId>,
+   // This vector is always empty when this struct is used for labels
+   labels: Vec<Box dyn SystemLabel>,
+}
+```
+
+### Flushed ordering constraints
+
+Being able to assert that commands are processed (or state is flushed) between two systems is a critical requirement for being able to specify logically-meaningful ordering of systems without a global view of the schedule.
+
+However, the exact logic involved in doing this is quite complex.
+We need to:
+
+1. **When validating the schedule:** ensure that a flushing system of the appropriate type *could* occur between the pair of systems.
+2. **When executing the schedule:** ensure that a flushing system of the appropriate type *does* occur between the pair of systems.
+
+For this RFC, let's begin with the very simplest strategy: for each pair of systems with a flushed ordering constraint, ensure that a flushing system of the appropriate type is strictly after the first system, and strictly before the second.
+
+We can check if the flushing system is of the appropriate type by checking the type id of the function pointer stored in the `RawSystem`.
+This is cached in the `system_function_type_map` field of the `Schedule` for efficient lookup, yielding the relevant `SystemIds`.
+
+With the ids of the relevant systems in hand, we can check each constraint: for each pair of systems, we must check that at least one of the relevant flushing systems is between them.
+To do so, we need to check both direct *and* transitive strict ordering dependencies, using a depth-first search to walk backwards from the later system using the `CachedOrdering` until a flushing system is reached, and then walking back further from that flushing system until the originating system is reached.
+
+## Drawbacks
+
+1. This will be a massively breaking change to every single Bevy user.
+   1. Any non-trivial control over system ordering will need to be completely reworked.
+   2. Users will typically need to think a bit harder about exactly when they want their gameplay systems to run. In most cases, they should just add the `CoreLabel::GameLogic` label to them.
+2. It will be harder to immediately understand the global structure of Bevy apps.
+   1. Powerful system debugging and visualization tools become even more important.
+3. State transitions are no longer queued up in a stack.
+   1. Arbitrary chains of state transitions can no longer be processed in the same frame, due to the lack of a state stack abstraction.
+   2. This also removes "in-stack" and related system groups / logic.
+   3. This can be easily added later, or third-party plugins can create their own abstraction.

```

Comment:
I think the first bullet point could be reworded for clarity. Users can't do arbitrary transition chains in `flush_state`, but they could add an arbitrary number of `flush_state` instances.